### ISSUE TITLE
datatype-expansion memory problems

### ISSUE THREAD

**Issue by kevinrenskers** [raml-org/datatype-expansion#49]
Please see https://github.com/raml2html/raml2html/issues/326 for a description of the issue, but basically the `expandedForm` function has memory issues with certain RAML files.

Test file: [api.raml.zip](https://github.com/raml2html/raml2html/files/771316/api.raml.zip)

```
#%RAML 1.0
title: some API
baseUri: https://someserver.com/api
mediaType: application/json

types:
  SomeProfile:
    type: object
    properties:
      attr1: string?
      attr2: string?
      attr3: string?
      attr4: string?
      attr5: string?
      attr6: string?
      attr7: string?
      attr8: string?
      attr9: string?
      attr10: string?
      attr11: string?
      attr12: string?
      attr13: string?
      attr14: string?
      attr15: string?

  ManyProfiles:
    type: object
    properties:
      profiles?: SomeProfile

/api:
  get:
    responses:
      200:
        body:
          type: ManyProfiles

```

**Comment by rmihael** [comment]
Any news on this problem? It really hurts our pipeline :(

**Comment by kevinrenskers** [comment]
@sichvoge have you been able to look at this memory problem?

**Comment by sichvoge** [comment]
No, not yet. We will try to look at this asap.

**Comment by ForsakenHarmony** [comment]
the `canonicalForm` function has problems, because it generates a 60MB json file, `expandedForm` should be fine?

**Comment by sichvoge** [comment]
> because it generates a 60MB json file

Is this the resulting JSON output? 60MB sounds a lot =O

**Comment by ForsakenHarmony** [comment]
well it exands every one of the `string?`s into `anyOf [{type string}, {type nil}]`
and then canonical makes the `anyOf` into a root element, listing every single combinations of those
so `2^15 = 32768` times the `SomeProfile` object

**Comment by jstoiko** [comment]
@kevinrenskers: we've just released v0.2.0 of `datatype-expansion` which is a drop-in replacement written in pure JavaScript. Let me know if you still experience this issue. If not, I'll close this.

**Comment by trevorr** [comment]
@kevinrenskers: Once #83 is merged and released, I'll submit a PR to raml2obj to disable union hoisting. I don't think hoisting unions is ever the desired behavior for documentation generation.
